### ISSUE TITLE
New lint: `split(" ")` where `split_whitespace()` was likely meant

### ISSUE THREAD

**Issue by omarmbyte** [rust-lang/rust-clippy#17693]
### What it does

Warns when `str::split(" ")` (or `split(' ')`) is used in a way that suggests the author wanted to split on whitespace generally, not on a literal single space. `split_whitespace()` handles multiple consecutive spaces, tabs, and leading/trailing whitespace correctly, while `split(" ")` produces empty strings between repeated spaces.

Example:
```rust
    println!("{:?}", "a b".split(" ").collect::<Vec<_>>());
    println!("{:?}", "a b".split_whitespace().collect::<Vec<_>>());
    println!("{:?}", "a  b".split(" ").collect::<Vec<_>>());
```
Output:
```
["a", "b"]
["a", "b"]
["a", "", "b"]
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&code=fn+main%28%29+%7B%0A++++println%21%28%22%7B%3A%3F%7D%22%2C+%22a+b%22.split%28%22+%22%29.collect%3A%3A%3CVec%3C_%3E%3E%28%29%29%3B%0A++++println%21%28%22%7B%3A%3F%7D%22%2C+%22a+b%22.split_whitespace%28%29.collect%3A%3A%3CVec%3C_%3E%3E%28%29%29%3B%0A++++println%21%28%22%7B%3A%3F%7D%22%2C+%22a++b%22.split%28%22+%22%29.collect%3A%3A%3CVec%3C_%3E%3E%28%29%29%3B%0A%0A%7D)

### Advantage

- Catches a bug that's easy to miss, `split(" ")` ignores extra spaces
- Points people toward the method that actually does what is probably meant

### Drawbacks

`split(" ")` and `split_whitespace() `don't do the same thing, so this can't be auto fixed, that would break anything relying on the empty strings, needs some thought on when it should trigger, like when the result is immediately filtered with `.filter(|p| !p.is_empty())` or warn without auto fixing.

### Example

```rust
let parts: Vec<&str> = input.split(" ").filter(|p| !p.is_empty()).collect();
```

Could be written as:

```rust
let parts: Vec<&str> = input.split_whitespace().collect();
```

### Comparison with existing lints

Similar to `trim_split_whitespace` which flags the unnecessary `.trim()` before `.split_whitespace()`, but this is a different case, since it's about `.trim()`, not `.split(" ")`.

There's also `non_portable_line_iteration` which suggested `.lines()` instead of `split("\n")` or `.split("\r\n")`.

**Comment by omarmbyte** [comment]
GitHub code search shows `.split(" ")` / `.split(' ')` appearing across ~101k Rust files

regex:
```regex
/\.split\((" "|' ')\)/
```

Narrowing to cases where a `.filter(...)` call follows shortly after (a sign someone is manually working around the empty string issue) still shows ~5.6k Rust files

regex: 
```regex
/\.split\((" "|' ')\)\s*\n?\s*\.filter\([^)]*\)/
```

**Comment by Szizoid** [comment]
I'd like to take this on

There are a few things I'd like to agree on first

1. The transformation is not equivalence-preserving

You can't simply replace `split(" ")`, or even `split(" ").filter(|s| !s.is_empty())`, with `split_whitespace()`. The behaviour differs, at the very least in that `split_whitespace()` also splits the string on other Unicode whitespace ("\t", "\n", and so on), so none of these forms is interchangeable with it:

```rust
let s = "a  \t  b";

s.split(" ").collect::<Vec<_>>();                            // ["a", "", "\t", "", "b"]
s.split(' ').collect::<Vec<_>>();                            // ["a", "", "\t", "", "b"]
s.split(" ").filter(|p| !p.is_empty()).collect::<Vec<_>>();  // ["a", "\t", "b"]

s.split_whitespace().collect::<Vec<_>>();                    // ["a", "b"]
```

The filter does get rid of the empty elements, but the tab still ends up as one of its own

I'd emit this suggestion as `MaybeIncorrect`, and would explicitly tell the user that the behaviour changes too: other whitespace characters will now be split on as well

2. When it should trigger

I suggest picking up only the cases where the user has shown the intent to get rid of the empty elements themselves:

```rust
some.split(" ").filter(|s| !s.is_empty())
some.split(" ").filter(|s| s.len() > 0)
some.split(" ").filter(|s| !s.trim().is_empty())
// and the same variants with `.split(' ')`
```

I also suggest deliberately ignoring the following situations:
- `split(" ")` with no `filter()` after it. There is a whole set of cases where this construct makes perfect sense: HTTP request lines, files in `/proc`, space-separated records
- `splitn(n, " ")`. Empty elements are often intentional here
- `split("  ")`, and any other number of spaces and/or other Unicode whitespace characters. The intent here is usually different, and usually deliberate as well

Support for indirect chains between `split()` and `filter()` (an intermediate `map(str::trim)`, for example) could be added to the implementation later

3. Which category to assign to the lint

The suggestion is opinionated, it changes semantics, and the pattern may not always be unambiguous enough to warn by default, so I suggest `pedantic`

**Comment by CommanderStorm** [comment]
> I'd emit this suggestion as `MaybeIncorrect`, and would explicitly tell the user that the behaviour changes too

similar to `str_split_at_newline`, yes.

> I suggest picking up only the cases where the user has shown the intent to get rid of the empty elements themselves

Makes sense for the first version, but a lot of code is written to not account multiple whitespaces, so which way this should work (lint exception or not) is maybe something for zulip.
Anyhow, removing the filter-check is simple to do anyhow if during the FCP this is found to be acceptable, which I don't know.

> I suggest pedantic

`str_split_at_newline` is also `pedantic`, makes sense.
This is one of the things that may be changed in the FCP anyhow.
